File: cpp/LeetCode/No487.cpp
/*
Given a binary array nums, return the maximum number of consecutive 1's in the array if you can flip at most one 0.
*/
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

class Solution {
public:
    int findMaxConsecutiveOnes(vector<int>& nums) {
        int res = 0;
        dp(nums, 0, false, 0, res);
        return res;
    }
private:
    void dp(vector<int> &nums, int index, bool flip, int count, int &res){
        if(index==nums.size() || (flip && nums[index]==0)){
            res = max(res, count);
            return;
        }

        if(nums[index]==0){
            dp(nums, index+1, true, count+1, res);
            dp(nums, index+1, flip, 0, res);
        }
        else{
            dp(nums, index+1, flip, count+1, res);
        }
    }
};

int main(void){
    Solution solution;
    vector<int> nums;

    nums = {1,0,1,1,0};
    cout << solution.findMaxConsecutiveOnes(nums) << endl;

    nums = {1,0,1,1,0,1};
    cout << solution.findMaxConsecutiveOnes(nums) << endl;

    return 0;
}

File: python/LeetCode Quiz/No487.py
'''
Given a binary array nums
, return the maximum number of consecutive 1's in the array if you can flip at most one 0.
'''
from typing import List

class Solution:
    def findMaxConsecutiveOnes(self, nums: List[int]) -> int:
        if len(nums)==1:
            return 1
        temp = [nums[0]]
        for i in range(1, len(nums)):
            if nums[i] == 1:
                if temp[-1]>0:
                    temp[-1]+=1
                else:
                    temp.append(1)
            else:
                temp.append(0)

        if len(temp)==1:
            return temp[0]
        res = 0
        for i, val in enumerate(temp):
            if val == 0:
                if i==0:
                    res = max(res, temp[1]+1)
                elif i==len(temp)-1:
                    res = max(res, temp[-2]+1)
                else:
                    res = max(res, temp[i-1] + temp[i+1] + 1)
        return res


def main():
    nums = [1,1]
    print(Solution.findMaxConsecutiveOnes(Solution(), nums))

if __name__=='__main__':
    main()